src/docs/binding-closures.md

```diff
@@ -0,0 +1,135 @@
+# Binding Closures
+
+Closures in Swift come in two general forms: escaping and non-escaping. A closure is considered escaping if it consinues to exist outside of the context that created it. For example, a closure paramater to a function that gets stored in a member variable in a class *must* be marked `@escaping` or the compiler will flag it as an error.
+
+In either case, Swift closures, unlike some other languages, are full-fledged closures that can have both free and bound variables within the body of the closure.
+
+The distinction between the two is that escaping closures that capture free variables use a dynamically allocated object to contain any captured variables. This object is reference counted. Non-escaping closures use stack allocated memory to contain any captured variables.
+
+Beyond the broad classification of escaping/non-escaping, closures can also be async and they can throw.
+
+Internally, closures are represented as 2 machine words:
+- Pointer to a function entry point for the closure (see below for details)
+- Pointer to the data context object for captured free variables or 0/null if there is no context object
+
+## Language Parity Mismatches
+
+None in particular. From a language standpoint, they both align in capability although the implementatioa are different.
+
+Swift has a delightful bit of syntax sugar that allows you to supply a closure at the call site in a way that it looks like inline code.
+If you have a function like this in Swift:
+```swift
+public func sorter<T>(arr: [T], by: (T, T) -> Int) { /* implementation not important */ }
+
+```
+And `sorter` can by called by any of the following means:
+
+```swift
+private func sort0 (a: Int, b: Int) -> Int {
+    return a - b
+}
+
+let arr = [7, 1, 27]
+sorter(arr: a, by: sort0) // 1: sort0 is your closure
+sorter(arr: a, by: { a, b in return a - b}) // 2: inline closure
+sorter(arr: a) { a, b in  // 3: trailing closure
+    return a - b
+}
+```
+The last case is a trailing closure which is exactly equivalent to the second inline closure
+
+## ABI Differences
+
+Closures follow the same calling conventions as [functions](binding-functions.md).
+
+With support for Swift calling conventions in the runtime, we should be insulated from issues in the ABI differences.
+
+We would need to be aware reference counting of the context object. Closures are more or like like this:
+```swift
+public struct EscapingClosure {
```

Comment:
Why not `class` since it is ref-counted?

Reply:
This is a Swift representation, not a C# representation. The closure itself isn't reference counted AFAIK only the `AnyObject` field